File: Calorimeter/CaloClusterCorrection/src/CaloSwEtamod_v2.cxx
/*
  Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration
*/
/**
 * @file  CaloSwEtamod_v2.cxx
 * @author scott snyder <snyder@bnl.gov>
 * @date Oct, 2004
 * @brief EM calorimeter @f$\eta@f$ modulation corrections.
 * 
 * Derived from previous versions:
 * AUTHORS:  H. Ma, S. Rajagopalan
 * CREATED:  Dec. 15, 1999
 * 
 * PURPOSE:  correction for the modulations with the position in eta
 *           (Tuned using 50 GeV Et photons)  
 *           base class: CaloClusterCorrectionCommon
 * 
 * Atrecon Orig: emreco/qetamod.F
 * 
 * Updated:  May 10, 2000    (SR, HM)
 *           Migrated to Athena Framework from PASO
 * 
 * Updated:  Jan 5, 2001    (HM)
 *           QA. 
 * 
 * Updated:  May 5, 2004    (Sven Menke)
 * 	  base class changed from algo to tool
 * 
 * Updated:  June, 2004    (sss)
 *           Use ToolWithConstants to get correction constants.
 */


#include "CaloSwEtamod_v2.h"
#include "CaloClusterCorrection/interpolate.h"
#include "CaloDetDescr/CaloDetDescrManager.h"
#include <cmath>


using xAOD::CaloCluster;
using CaloClusterCorr::interpolate;
using std::abs;


/**
 * @brief Virtual function for the correction-specific code.
 * @param myctx   ToolWithConstants context.
 * @param cluster The cluster to correct.
 *                It is updated in place.
 * @param elt     The detector description element corresponding
 *                to the cluster location.
 * @param eta     The @f$\eta@f$ coordinate of the cluster, in this sampling.
 * @param adj_eta The @f$\eta@f$ adjusted for
 *                any shift between the actual and nominal coordinates.
 *                (This is shifted back to the nominal coordinate system.)
 * @param phi     The @f$\phi@f$ coordinate of the cluster, in this sampling.
 * @param adj_phi The @f$\phi@f$ adjusted for
 *                any shift between the actual and nominal coordinates.
 *                (This is shifted back to the nominal coordinate system.)
 * @param samp    The calorimeter sampling we're examining.
 *                This is a sampling code as defined by
 *                @c CaloSampling::CaloSample; i.e., it has both
 *                the calorimeter region and sampling encoded.
 */
void CaloSwEtamod_v2::makeTheCorrection (const Context& myctx,
                                         CaloCluster* cluster,
                                         const CaloDetDescrElement* elt,
                                         float eta,
                                         float adj_eta,
                                         float /*phi*/,
                                         float /*adj_phi*/,
                                         CaloSampling::CaloSample /*samp*/)
  const
{
  // Compute the eta offset within the cell.
  float etamod = eta - elt->eta() + elt->deta()/2;
  if (elt->eta_raw() < 0)
    etamod = elt->deta() - etamod;

  // These can happen due to DD bugs.
  if (etamod < 0)
    etamod = 0;
  else if (etamod > elt->deta())
    etamod = elt->deta();

  float adj_aeta = std::abs (adj_eta);

  // ??? In most of the calorimeter, the cells in sampling 2 have an eta
  //     width of 0.025.  This correction was derived with this assumption.
  //     However, it's not completely accurate: in the gap region, some
  //     cells have a different size.  For example, the last sampling 2
  //     cell in the barrel (starting at 1.4) has a width of 0.075.
  //     For this pass, fix up etamod to match what was used when
  //     the correction was derived.  This should be cleaned up in
  //     a subsequent version.
  if (adj_aeta >= 1.4 && adj_aeta <= 1.475)
    etamod = fmod (adj_aeta, 0.025);

  // Before doing the energy interpolation, make a crude total correction
  // of the energy.  This is needed since the corrections are tabulated
  // using the true cluster energies.
  float energy = cluster->e();
  float rfac = interpolate (m_rfac(myctx), adj_aeta, m_rfac_degree(myctx));
  energy /= rfac;

  // Calculate the correction.
  float corr = energy_interpolation (energy,
                                     Builder (m_correction(myctx), etamod),
                                     m_energies(myctx),
                                     m_energy_degree(myctx));

  // set energy, and rescale each sampling
  setenergy (cluster, cluster->e() / corr);
}


/**
 * @brief Constructor for energy interpolation table helper class.
 * @param correction The array of correction parameters.
 * @param etamod The eta offset within the cell.
 */
CaloSwEtamod_v2::Builder::Builder (const CxxUtils::Array<2>& correction,
                                   float etamod)
  : m_correction (correction),
    m_etamod (etamod)
{
}


/**
 * @brief Calculate the correction for one tabulated energy index.
 * @param energy_ndx The tabulated energy index for which
 *                   the correction should be calculated.
 * @param good[out]  Set to true if the calculation succeeds.
 */
float CaloSwEtamod_v2::Builder::calculate (int energy_ndx, bool& good) const
{
  good = true;
  return m_correction[energy_ndx][0] 
    + m_etamod*m_correction[energy_ndx][1]
    + m_etamod*m_etamod*m_correction[energy_ndx][2];
}

File: Calorimeter/CaloClusterCorrection/python/CaloSwEtamod_v2.py
# Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration

#
# File: CaloClusterCorrection/python/CaloSwEtamod_v2.py
# Created: Nov 2006, sss
# Purpose: Eta modulation corrections, v2.
#
# These corrections were derived using DC2 MC files reconstructed
# with 8.x.0.  Electrons at 50, 100, and 200 GeV were used.
# Derived for 5x5 clusters only.
#
# This was added in LArClusterRec-02-05-25, in 9.0.3.
#
# Two versions of this correction are exported.
# The original (v2) version of this correction ran before the
# gross energy correction.  Therefore, when the energy interpolation
# was performed, the energy was scaled with a crude eta-dependent
# factor.  For v3 of the corrections, the modulation corrections
# were moved to after the gross energy corrections.  Accordingly,
# this crude energy correction was turned off (by setting the
# correction factor to unity).
#
# v3 of the correction was in CaloClusterCorrection-00-02-58, in 12.0.3.
#


from CaloClusterCorrection.constants import CALOCORR_COMBINED2
from AthenaCommon.SystemOfUnits import GeV
from CaloClusterCorrection.CaloSwEtaphimod_v2_rfac import CaloSwEtaphimod_v2_rfac


# Energies at which corrections were tabulated.
CaloSw_sample_energies = [50*GeV, 100*GeV, 200*GeV]


etamod_v3_correction = [
        [0.9988, 0.3031, -12.0],
        [0.9984, 0.356,  -13.55],
        [0.9987, 0.2368, -7.566]
    ]


# Version which goes before the gross energy corrections.
class CaloSwEtamod_v2_parms:
    region = CALOCORR_COMBINED2
    energies = CaloSw_sample_energies
    energy_degree = 2
    rfac = CaloSwEtaphimod_v2_rfac
    rfac_degree = 2
    correction = etamod_v3_correction


# Version which goes after the gross energy corrections.
class CaloSwEtamod_v3_parms:
    region = CALOCORR_COMBINED2
    energies = CaloSw_sample_energies
    energy_degree = 2
    rfac = [ [0, 1], [3, 1] ]
    rfac_degree = 1
    correction = etamod_v3_correction
